File: aoc2025-go/days/day03/day03.go
package day03

import (
	"fmt"
	"strconv"
	"time"

	"aoc2025-go/common"
	"aoc2025-go/parse"
)

func getValueOfDigit(voltage byte) int {
	return int(voltage - ZERO)
}

const ZERO = 48

func pickDigit(bank string, chosen_digits string, start int, batteries int) string {
	n := len(chosen_digits)
	if n == batteries {
		return chosen_digits
	}
	candidate := byte('0')
	index_of_candidate := start

	for i := 0; i <= len(bank)-(batteries-n); i++ {
		val := byte(bank[i])
		if getValueOfDigit(val) > getValueOfDigit(candidate) {
			candidate = val
			index_of_candidate = i
		}
	}

	return pickDigit(bank[index_of_candidate+1:], chosen_digits+string(candidate), index_of_candidate, batteries)

}

func Part1(banks []string) string {
	var sum int
	for _, bank := range banks {
		chosen_digits := pickDigit(bank, "", 0, 2)
		value, _ := strconv.Atoi(chosen_digits)
		sum += value
	}
	return fmt.Sprintf("%d", sum)
}

func Part2(banks []string) string {
	var sum int
	for _, bank := range banks {
		chosen_digits := pickDigit(bank, "", 0, 12)
		value, _ := strconv.Atoi(chosen_digits)
		sum += value
	}

	return fmt.Sprintf("%d", sum)
}

func Run() {
	input := parse.Lines(3, false)

	t1s := time.Now()
	p1 := Part1(input)
	t1 := time.Since(t1s)

	t2s := time.Now()
	p2 := Part2(input)
	t2 := time.Since(t2s)

	common.RunDay(3, p1, p2, t1, t2)
}

File: aoc2025-zig/src/days/day03.zig
const std = @import("std");
const common = @import("../common.zig");
const parse = @import("../parse.zig");

fn pickDigits(bank: []const u8, digits: []const u8, start: usize, batteries: usize, allocator: std.mem.Allocator) ![]const u8 {
    const n = digits.len;
    if (n == batteries) return digits;

    var candidate: u8 = '0';
    var index_of_candidate: usize = start;

    for (0..bank.len - (batteries - n) + 1) |i| {
        const val = bank[i];
        if (val > candidate) {
            candidate = val;
            index_of_candidate = i;
        }
    }

    const combined = try std.mem.concat(allocator, u8, &.{ digits, &[_]u8{candidate} });
    return pickDigits(bank[index_of_candidate + 1 ..], combined, index_of_candidate, batteries, allocator);
}

pub fn part1(banks: [][]const u8, allocator: std.mem.Allocator) ![]const u8 {
    var sum: usize = 0;
    for (banks) |bank| {
        const digits = try pickDigits(bank, "", 0, 2, allocator);
        sum += try std.fmt.parseInt(usize, digits, 10);
    }
    return std.fmt.allocPrint(allocator, "{d}", .{sum});
}

pub fn part2(banks: [][]const u8, allocator: std.mem.Allocator) ![]const u8 {
    var sum: usize = 0;
    for (banks) |bank| {
        const digits = try pickDigits(bank, "", 0, 12, allocator);
        sum += try std.fmt.parseInt(usize, digits, 10);
    }
    return std.fmt.allocPrint(allocator, "{d}", .{sum});
}

pub fn run(io: std.Io, allocator: std.mem.Allocator) !void {
    const input = try parse.lines(io, allocator, 3, false);

    const t1s = std.Io.Timestamp.now(io, .awake);
    const p1 = try part1(input, allocator);
    const t1e = std.Io.Timestamp.now(io, .awake);

    const t2s = std.Io.Timestamp.now(io, .awake);
    const p2 = try part2(input, allocator);
    const t2e = std.Io.Timestamp.now(io, .awake);

    common.runDay(io, 3, p1, p2, t1s, t1e, t2s, t2e);
}
